### PR TITLE
Namespace diagnostic fixes

### PR DESCRIPTION
**Pull request by j0shuams**
This PR addresses two diagnostic issues with namespaces #732 and #733 

I am getting rid of the "TypeCollector" I used to check for the "no public types" diagnostic, and instead checking in the SyntaxReceiver. If a namespace declaration has no public types, we don't add it as a declaration; then on the diagnostic side if we have no declarations, we raise the "no public types" diagnostic. 

I also changed the way we check namespace names for conflicts -- this is because we cannot allow namespaces like "Sample" and "SaMpLe, which only differ by case. We are getting rid of the Contains method and instead looping through the set of namespace names seen so far -- a bit of a pain to not have O(1) method, but I didn't think of another way.

Also making some quality of life/code adjustments like formatting and a readme for adding unit tests.

### CODE REVIEW

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 1]
```diff
@@ -222,9 +224,20 @@ class WinRTSyntaxReciever : ISyntaxReceiver
     {
         public List<MemberDeclarationSyntax> Declarations = new List<MemberDeclarationSyntax>();
 
+        private bool HasSomePublicTypes(SyntaxNode syntaxNode)
+        {
+            bool hasSomePublicTypes = false;
```
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:229]: This probably should include enums too and delegates?  Should we just look for any public child node?

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 2]
```diff
@@ -203,10 +203,12 @@ public class SourceGenerator : ISourceGenerator
     {
         public void Execute(GeneratorExecutionContext context)
         {
+            /*
```
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:206]: intentional?

Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/Generator.cs:206]: Yes, it's temporary for testing, will be removed in final commit.

File: src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs [hunk 3]
```diff
@@ -15,25 +16,31 @@ public WinRTComponentScanner(GeneratorExecutionContext context, string assemblyN
...
-            HasSomePublicTypes();
-
+        {
             WinRTSyntaxReciever syntaxReciever = (WinRTSyntaxReciever)_context.SyntaxReceiver;
+
+            if (!syntaxReciever.Declarations.Any())
+            {
+                Report(WinRTRules.NoPublicTypesRule, null);
+                // exit early?
+                // return;
```
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/DiagnosticUtils.cs:38]: Anything to do if no types?

File: src/Authoring/WinRT.SourceGenerator/Generator.cs (hunk 1 above)
Comment (j0shuams) [reply src/Authoring/WinRT.SourceGenerator/Generator.cs]: Updated, seems to work fine.

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 4]
```diff
@@ -222,9 +224,16 @@ class WinRTSyntaxReciever : ISyntaxReceiver
...
+        {
+            return syntaxNode.ChildNodes().OfType<MemberDeclarationSyntax>().Any(IsPublic);
+        }
+
         public void OnVisitSyntaxNode(SyntaxNode syntaxNode)
         {
-            if (syntaxNode is not MemberDeclarationSyntax decaralation || !IsPublic(decaralation))
+            // We will lose namespaces in `Declarations` if we don't special case them as exempt from the `IsPublic` check
+            if (syntaxNode is not MemberDeclarationSyntax decaralation || 
+                (syntaxNode is not NamespaceDeclarationSyntax @namespace && !IsPublic(decaralation)))
```
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:236]: probably can remove @namespace as we don't use it.

**Review by manodasanW — APPROVED** [review]

File: src/Authoring/WinRT.SourceGenerator/Generator.cs [hunk 5]
```diff
@@ -231,9 +232,23 @@ private bool HasSomePublicTypes(SyntaxNode syntaxNode)
...
-            // We will lose namespaces in `Declarations` if we don't special case them as exempt from the `IsPublic` check
-            if (syntaxNode is not MemberDeclarationSyntax decaralation || 
-                (syntaxNode is not NamespaceDeclarationSyntax @namespace && !IsPublic(decaralation)))
+            // Store namespaces separately as we only need to look at them for diagnostics
+            // If we did store them in declarations, we would get duplicate entries in the WinMD,
+            // once from the namespace declaration and once from the member's declaration
+            if (syntaxNode is NamespaceDeclarationSyntax @namespace)
+            {
+                if (!HasSomePublicTypes(syntaxNode))
+                {
```
Comment (manodasanW) [inline src/Authoring/WinRT.SourceGenerator/Generator.cs:241]: We might want to just return always if it enters this if namespace as it is not going to meet the checks after.

**Pull request merged**